network_simulator/host.py

```diff
@@ -23,4 +27,65 @@ def __init__(self, address=None, flows={}, flow_packets_received={},
         self.flow_packets_received = flow_packets_received
         self.link = link
 
-    # TODO(team): More event subclasses to come?
\ No newline at end of file
+
+class HostReceivedPacketEvent(Event):
+    """
+    Host receives packets from a flow.
+    """
+
+    def __init__(self, host, packet):
+        """
+        Sanity check: make sure destination was this Host and that this isn't
+        a RouterPacket.
+        :ivar host: host that receives packet
+        :ivar packet: packet received
+        """
+        Event.__init__(self)
+
+        # Sanity check
+        assert packet.dest_id == host and \
+        packet.packet_id != ROUTER_PACKET_DEFAULT_ID
+
+        self.host = host
+        self.packet = packet
+
+    def run(self, main_event_loop, statistics):
+        """
+        Put packet in flow_packets_received and sort.
+        Update statistics.
+        :param main_event_loop: event loop where new Events will be scheduled.
+        :param statistics: statistics to update
+        """
+        # add packet to host.flow_packets_received
+        self.flow_packets_received[self.packet.flow_id] = self.packet.packet_id
+        self.flow_packets_received.sort()
+        
+        statistics.host_packet_received(self.host, self.packet,
+                                        main_event_loop.global_clock_sec)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        If packet was Ack, look up Flow and say packet was received.
+        Schedule FlowReceivedAckEvent.
+        Populate AckPacket for selective repeat.
+
+        If packet was data packet, generate Ack and schedule via
+        DeviceToLink event unless packet already received earlier.
+
+        Ackpackets, set original data packet's start time for RTT.
+        AckPacket ID same as DataPacket ID
+        :param main_event_loop: schedule new Events
+        :param statistics: update statistics
+        """
+
+        if isinstance(self.packet, AckPacket):
+            main_event_loop.schedule_new_event_with_delay(self, self.flow,
```

Comment:
This section does not make sense since this is not the proper way to add Events to the MainEventLoop. What you need to do is the following:

1) Look up the Flow corresponding to this AckPacket. We already know the flow_id; it is in self.packet.flow_id. What you need to do is look up self.flows[<the flow ID>], to get the Flow starting at this Host. Because self.flows is a "map from flow ID to Flows starting at the Host".
2) Construct a FlowReceivedAckEvent object with the flow you looked up and the received packet. See flow.py for the constructor signature for this Event.
3) Add the FlowReceivedAckEvent object you made to the Event loop. This is done via a call of the form main_event_loop.schedule_event_with_delay(<your constructed Event>, 0.0)

So all you are doing in this case is relaying the ACK onto the Flow, which will handle it.